beetle-core/src/main/java/com/xing/beetle/dedup/spi/Deduplicator.java

```diff
@@ -57,13 +56,17 @@
 
   default <M> void runHandler(
       M message, MessageListener<M> listener, MessageAdapter<M> adapter, Duration timeout) {
-    Interruptable<M> interruptable = new Interruptable<>(listener);
-    // Schedule an interruption for the execution of the handler when the timeout is expired
-    CompletableFuture.delayedExecutor(timeout.toMillis(), TimeUnit.MILLISECONDS)
-        .execute(interruptable::interruptTimedOutAndRethrow);
-    // actually run the handler, i.e handle the message
     try {
-      interruptable.onMessage(message);
+      CompletableFuture<Void> cf =
+          CompletableFuture.runAsync(
+              () -> {
+                try {
+                  listener.onMessage(message);
+                } catch (Throwable t) {
+                  ExceptionSupport.sneakyThrow(t);
+                }
+              });
+      cf.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
     } catch (Throwable throwable) {
       if (throwable instanceof InterruptedException
```

Comment:
This part needs to be changed as well, since we do not depend on InterruptedException to detect timouts. We should be looking for TimeoutExceptions here, since this line will throw when a timeout happens: `cf.get(timeout.toMillis(), TimeUnit.MILLISECONDS);`